include/cuco/static_multimap.cuh

```diff
@@ -0,0 +1,1716 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cooperative_groups.h>
+#include <thrust/distance.h>
+#include <thrust/functional.h>
+#include <cub/cub.cuh>
+#include <memory>
+
+#include <cuco/allocator.hpp>
+
+#if defined(CUDART_VERSION) && (CUDART_VERSION >= 11000) && defined(__CUDA_ARCH__) && \
+  (__CUDA_ARCH__ >= 700)
+#define CUCO_HAS_CUDA_BARRIER
+#endif
+
+// cg::memcpy_aysnc is supported for CUDA 11.1 and up
+#if defined(CUDART_VERSION) && (CUDART_VERSION >= 11100)
+#define SUPPORTS_CG_MEMCPY_ASYNC 1
+#else
+#define SUPPORTS_CG_MEMCPY_ASYNC 0
+#endif
+
+#if defined(CUCO_HAS_CUDA_BARRIER)
+#include <cuda/barrier>
+#endif
+
+#include <cuco/detail/bitwise_compare.cuh>
+#include <cuco/detail/error.hpp>
+#include <cuco/detail/prime.hpp>
+#include <cuco/detail/probe_sequences.cuh>
+#include <cuco/detail/static_multimap_kernels.cuh>
+
+namespace cuco {
+
+/**
+ * @brief A GPU-accelerated, unordered, associative container of key-value
+ * pairs that supports equivalent keys.
+ *
+ * Allows constant time concurrent inserts or concurrent find operations (not
+ * concurrent insert and find) from threads in device code.
+ *
+ * Current limitations:
+ * - Requires keys and values that where `cuco::detail::is_bitwise_comparable<T>::value` is true
+ * - Comparisons against the "sentinel" values will always be done with bitwise comparisons
+ * Therefore, the objects must have unique, bitwise object representations (e.g., no padding bits).
+ * - Does not support erasing keys
+ * - Capacity is fixed and will not grow automatically
+ * - Requires the user to specify sentinel values for both key and mapped value
+ * to indicate empty slots
+ * - Does not support concurrent insert and find operations
+ *
+ * The `static_multimap` supports two types of operations:
+ * - Host-side "bulk" operations
+ * - Device-side "singular" operations
+ *
+ * The host-side bulk operations include `insert`, `contains`, `count`, `retrieve` and their
+ * variants. These APIs should be used when there are a large number of keys to insert or lookup in
+ * the map. For example, given a range of keys specified by device-accessible iterators, the bulk
+ * `insert` function will insert all keys into the map.
+ *
+ * The singular device-side operations allow individual threads to perform
+ * independent operations (e.g. `insert`, etc.) from device code. These
+ * operations are accessed through non-owning, trivially copyable "view" types:
+ * `device_view` and `device_mutable_view`. The `device_view` class is an
+ * immutable view that allows only non-modifying operations such as `count` or
+ * `contains`. The `device_mutable_view` class only allows `insert` operations.
+ * The two types are separate to prevent erroneous concurrent insert/find
+ * operations.
+ *
```

Comment:
In these top-level docs, we should have a description of the `pair_` prefix and `_outer` suffix.